cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -516,42 +520,39 @@ private void beforeSubmit(final Job job) throws Exception {
       : TransactionControl.IMPLICIT;
 
     if (TransactionControl.EXPLICIT == txControl) {
-      ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(job.getConfiguration().getClassLoader());
-      try {
-        doInitialize();
-      } finally {
-        ClassLoaders.setContextClassLoader(oldClassLoader);
-      }
-    }
-    TransactionContext txContext = context.getTransactionContext();
-    Transactions.execute(txContext, "initialize", new Callable<Void>() {
-      @Override
-      public Void call() throws Exception {
-        ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(job.getConfiguration().getClassLoader());
-        try {
-          if (TransactionControl.IMPLICIT == txControl) {
-            doInitialize();
-          }
-          // set input/outputs info, and get one of the configured mapper's TypeToken
-          TypeToken<?> mapperTypeToken = setInputsIfNeeded(job);
-          setOutputsIfNeeded(job);
-          setOutputClassesIfNeeded(job, mapperTypeToken);
-          setMapOutputClassesIfNeeded(job, mapperTypeToken);
-          return null;
-        } finally {
-          ClassLoaders.setContextClassLoader(oldClassLoader);
```

Comment:
change here is: 
- the class loader switching is moved to doInitialize()
- the setup after doInitialize() (setInputsIfNeeded, setMapOutputClassesIfNeeded, etc.) do not need a transaction any more, because they do not call dataset methods.